File: problems/007/007.lua
posix = require "posix"

local function f()
    local i, map, n, q = 0, {}, 10000, 1

    while true do
        q = q + 1

        p = map[q]

        if p == nil then
            map[q * q] = q

            if i == n then
                return q
            else
                i = i + 1
            end
        else
            map[q] = nil
            x = p + q

            while map[x] ~= nil do
                x = x + p
            end

            map[x] = p
        end
    end
end

function toNs(ts)
    return ts[1] * 1000000000 + ts[2]
end

if #arg == 0 then
    print(f())
elseif #arg == 1 then
    iters = tonumber(arg[1])
    start = { posix.clock_gettime("monotonic") }
    for i = 1,iters do
        f()
    end
    end_ = { posix.clock_gettime("monotonic") }
    print(toNs(end_) - toNs(start))
end

File: problems/007/007.js
function f() {
    var map = [];
    var i = 0;
    var n = 10000;
    var q = 1;
    var p, x;

    while (true) {
        q += 1;

        p = map[q];

        if (p === undefined) {
            map[q * q] = q;

            if (i === n) {
                return q;
            }

            i += 1;
        } else {
            delete map[q];

            x = p + q;

            while (map[x] !== undefined) {
                x += p;
            }

            map[x] = p;
        }

    }
}

function toNs(ts) {
    return ts.sec * 1000000000 + ts.nsec;
}

var clock = require('posix-clock');

if (process.argv.length === 3) {
    var i;
    var iters = parseInt(process.argv[2], 10);

    var start = clock.gettime(clock.MONOTONIC);
    for (i = 0; i < iters; i++) {
        f();
    }
    var end = clock.gettime(clock.MONOTONIC);

    console.log(toNs(end) - toNs(start));
} else {
    console.log(f());
}
